File: desktop-mend/src/test/java/commands/LockTest.java
package commands;

import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.core.Settings;
import co.samco.mend4.desktop.commands.Lock;
import co.samco.mend4.desktop.helper.ShredHelper;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import testutils.TestUtils;

import java.io.File;
import java.io.IOException;
import java.util.Collections;

import static org.mockito.Mockito.*;

public class LockTest extends TestBase {
   private Lock lock;
   private String privateKeyPath = "privpath";
   private String publicKeyPath = "pubpath";

   @Before
   public void setup() {
      super.setup();
      shredHelper = new ShredHelper(strings, osDao, settings, log);
      when(fileResolveHelper.getPrivateKeyPath()).thenReturn(privateKeyPath);
      when(fileResolveHelper.getPublicKeyPath()).thenReturn(publicKeyPath);
      lock = new Lock(strings, log, osDao, shredHelper, fileResolveHelper);
   }

   private ArgumentCaptor<String> getOutputOfLockTest() throws IOException, CorruptSettingsException {
      when(settings.getValue(Settings.Name.SHREDCOMMAND)).thenReturn("");
      Process process = mock(Process.class);
      when(process.getInputStream()).thenReturn(TestUtils.getEmptyInputStream());
      ArgumentCaptor<String> stdErr = ArgumentCaptor.forClass(String.class);
      when(osDao.executeCommand(any(String[].class))).thenReturn(process);
      lock.execute(Collections.emptyList());
      return stdErr;
   }

   private void assertCleaning(ArgumentCaptor<String> stdErr, int startInd) {
      Assert.assertEquals(strings.getf("Shred.cleaning", privateKeyPath), stdErr.getAllValues().get(startInd));
      Assert.assertEquals(strings.getf("Shred.cleaning", publicKeyPath), stdErr.getAllValues().get(startInd + 1));
   }

   @Test
   public void testKeyNotFound() throws CorruptSettingsException, IOException {
      when(osDao.fileExists(any(File.class))).thenReturn(false);
      ArgumentCaptor<String> stdErr = getOutputOfLockTest();
      verify(err, times(4)).println(stdErr.capture());
      Assert.assertEquals(strings.get("Lock.notUnlocked"), stdErr.getAllValues().get(0));
      assertCleaning(stdErr, 1);
      Assert.assertEquals(strings.get("Lock.locked"), stdErr.getAllValues().get(3));
   }

   @Test
   public void testKeyFoundAndLockFailed() throws CorruptSettingsException, IOException {
      when(osDao.fileExists(any(File.class))).thenReturn(true);
      ArgumentCaptor<String> stdErr = getOutputOfLockTest();
      verify(err, times(3)).println(stdErr.capture());
      assertCleaning(stdErr, 0);
      Assert.assertEquals(strings.get("Lock.lockFailed"), stdErr.getAllValues().get(2));
   }

   @Test
   public void testKeyFoundAndLockPassed() throws CorruptSettingsException, IOException {
      when(osDao.fileExists(any(File.class))).thenAnswer(new Answer<Boolean>() {
         int count = 0;
         @Override
         public Boolean answer(InvocationOnMock invocation) {
            if (count++ > 0) return false;
            else return true;
         }
      });
      ArgumentCaptor<String> stdErr = getOutputOfLockTest();
      verify(err, times(3)).println(stdErr.capture());
      assertCleaning(stdErr, 0);
      Assert.assertEquals(strings.get("Lock.locked"), stdErr.getAllValues().get(2));
   }
}

File: desktop-mend/src/test/java/commands/LockTest.kt
package commands

import co.samco.mend4.desktop.commands.Lock
import co.samco.mend4.desktop.dao.SettingsDao
import co.samco.mend4.desktop.helper.impl.ShredHelperImpl
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import testutils.TestUtils
import java.io.File

class LockTest : TestBase() {
    private lateinit var lock: Lock

    private val privateKeyPath = "test-privpath"
    private val publicKeyPath = "test-pubpath"

    private val privateKeyFile = File(privateKeyPath)
    private val publicKeyFile = File(publicKeyPath)

    @Before
    override fun setup() {
        super.setup()
        whenever(fileResolveHelper.privateKeyFile).thenReturn(privateKeyFile)
        whenever(fileResolveHelper.publicKeyFile).thenReturn(publicKeyFile)

        val shredHelper = ShredHelperImpl(
            strings = strings,
            settings = settings,
            log = log,
            fileResolveHelper = fileResolveHelper,
            osDao = osDao,
        )

        lock = Lock(
            log = log,
            settingsHelper = settingsHelper,
            shredHelper = shredHelper,
            strings = strings,
            fileResolveHelper = fileResolveHelper,
            osDao = osDao
        )
    }

    @Test
    fun testCommandWithSettingsDependencies() {
        super.testCommandWithNoSettingsDependencies(lock)
    }

    private fun testLock() {
        whenever(settings.getValue(SettingsDao.SHRED_COMMAND)).thenReturn("")
        val process = mock<Process>()
        whenever(process.inputStream).thenReturn(TestUtils.emptyInputStream)
        whenever(process.errorStream).thenReturn(TestUtils.emptyInputStream)
        whenever(osDao.exec(any())).thenReturn(process)
        lock.execute(emptyList())
    }

    private fun assertCleaning(stdErr: KArgumentCaptor<String>, startInd: Int) {
        Assert.assertEquals(
            strings.getf("Shred.cleaning", privateKeyFile.absolutePath),
            stdErr.allValues[startInd]
        )
        Assert.assertEquals(
            strings.getf("Shred.cleaning", publicKeyFile.absolutePath),
            stdErr.allValues[startInd + 1]
        )
    }

    @Test
    fun testKeyNotFound() {
        whenever(osDao.exists(any())).thenReturn(false)
        val stdErr = argumentCaptor<String>()
        testLock()
        verify(err, times(4)).println(stdErr.capture())
        println(stdErr.allValues)
        Assert.assertEquals(strings["Lock.notUnlocked"], stdErr.allValues[0])
        assertCleaning(stdErr, 1)
        Assert.assertEquals(strings["Lock.locked"], stdErr.allValues[3])
    }

    @Test
    fun testKeyFoundAndLockFailed() {
        whenever(osDao.exists(any())).thenReturn(true)
        val stdErr = argumentCaptor<String>()
        testLock()
        verify(err, times(3)).println(stdErr.capture())
        assertCleaning(stdErr, 0)
        Assert.assertEquals(strings["Lock.lockFailed"], stdErr.allValues[2])
    }

    @Test
    fun testKeyFoundAndLockPassed() {
        var count = 0
        whenever(osDao.exists(any())).thenAnswer { count++ <= 0 }
        val stdErr = argumentCaptor<String>()
        testLock()
        verify(err, times(3)).println(stdErr.capture())
        assertCleaning(stdErr, 0)
        Assert.assertEquals(strings["Lock.locked"], stdErr.allValues[2])
    }
}